components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsStorage.kt

```diff
@@ -0,0 +1,75 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.sitepermissions
+
+import android.content.Context
+import android.support.annotation.VisibleForTesting
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.async
+import kotlinx.coroutines.launch
+import mozilla.components.feature.sitepermissions.db.SitePermissionsDatabase
+import mozilla.components.feature.sitepermissions.db.toSitePermissionsEntity
+
+/**
+ * A storage implementation to save [SitePermissions].
+ *
+ */
+class SitePermissionsStorage(
+    context: Context
+) {
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal var databaseInitializer = {
+        SitePermissionsDatabase.get(context)
+    }
+
+    private val database by lazy { databaseInitializer() }
+
+    internal fun save(sitePermissions: SitePermissions) {
+        CoroutineScope(Dispatchers.IO).launch {
+            database
+                .sitePermissionsDao()
+                .insertOrUpdateSitePermission(
+                    sitePermissions.toSitePermissionsEntity()
+                )
+        }
+    }
+
+    fun update(sitePermissions: SitePermissions) {
+        save(sitePermissions)
+    }
+
+    fun findSitePermissionsBy(origin: String): Deferred<SitePermissions?> {
+        return GlobalScope.async(Dispatchers.IO) {
+            database
+                .sitePermissionsDao()
+                .getSitePermissionsBy(origin)
+                ?.toSitePermission()
+        }
+    }
+
+    fun all(limit: Int = 20): Deferred<List<SitePermissions>> {
+        return GlobalScope.async(Dispatchers.IO) {
+            database
+                .sitePermissionsDao()
+                .getSitePermissions(limit)
+                .map {
+                    it.toSitePermission()
+                }
+        }
+    }
```

Comment:
I think I would make all those methods blocking and let the calling app decide if it wants to use coroutines or rxjava or whatever. This also allows an app  using coroutines to define the scope too. For example here the call to `all()` should probably be in the scope of the activity/fragment and not in the global one.

Reply:
👍